src/dawn/Optimizer/PassSetNonTempCaches.cpp

```diff
@@ -0,0 +1,325 @@
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/Optimizer/PassSetNonTempCaches.h"
+#include "dawn/Optimizer/Cache.h"
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Optimizer/PassDataLocalityMetric.h"
+#include "dawn/Optimizer/PassSetCaches.h"
+#include "dawn/Optimizer/StatementAccessesPair.h"
+#include "dawn/Optimizer/StencilInstantiation.h"
+#include "dawn/SIR/ASTExpr.h"
+#include "dawn/Support/Unreachable.h"
+#include <iostream>
+#include <set>
+#include <vector>
+
+namespace dawn {
+
+struct accessIDaccessMetric {
+  accessIDaccessMetric(int id, int gain) : accessID(id), dataLocalityGain(gain) {}
+  int accessID;
+  int dataLocalityGain;
+  bool operator<(const accessIDaccessMetric& rhs) {
+    return dataLocalityGain < rhs.dataLocalityGain;
+  }
+};
+
+struct nameToImprovementMetric {
+  nameToImprovementMetric(std::string name_, Cache cache_, int metric)
+      : name(name_), cache(cache_), dataLocalityImprovement(metric) {}
+  std::string name;
+  Cache cache;
+  int dataLocalityImprovement;
+};
+
+/// @brief The GlobalFieldCacher class handles the caching for a given Multistage
+/// @param[in, out]  msprt   Pointer to the multistage to handle
+/// @param[in, out]  si      Stencil Instanciation [ISIR] holding all the Stencils
+class GlobalFieldCacher {
+public:
+  GlobalFieldCacher(MultiStage* msptr, StencilInstantiation* si)
+      : multiStagePrt_(msptr), instantiation_(si) {}
+
+  /// @brief entry method for the pass: processes a given multistage and applies all changes
+  /// required
+  void process() {
+    computeOptimalFields();
+    addFillerStages();
+    addFlushStages();
+  }
+
+  std::vector<nameToImprovementMetric> getoriginalNameToCache() { return originalNameToCache_; }
+
+private:
+  /// @brief use the data locality metric to rank the fields in a stencil based on how much they
+  /// would benefit from caching
+  void computeOptimalFields() {
+    auto dataLocality =
+        computeReadWriteAccessesMetricIndividually(instantiation_, *multiStagePrt_, config_);
+
+    for(const auto& stagePtr : multiStagePrt_->getStages()) {
+      for(const Field& field : stagePtr->getFields()) {
+        auto findField = accessIDtoDataLocality_.find(field.AccessID);
+
+        // We already checked this field
+        if(findField != accessIDtoDataLocality_.end())
+          continue;
+
+        // Only ij-cache fields that are not horizontally pointwise and are vertically pointwise
+        if(!(field.Extent.isVerticalPointwise()) || (field.Extent.isHorizontalPointwise()))
```

Comment:
parenthesis are not needed here operator **!** takes precedence over **||**

Reply:
I was thinking of having them there for readability but can see the argument for removing them